frappe/public/js/frappe/form/print.js

```diff
@@ -326,3 +469,119 @@ frappe.ui.get_print_settings = function (pdf, callback, letter_head) {
 		callback(data);
 	}, __("Print Settings"));
 }
+
+
+// qz tray connection wrapper
+//  - allows active and inactive connections to resolve regardless
+//  - try to connect once before firing the mimetype launcher
+//  - if connection fails, catch the reject, fire the mimetype launcher
+//  - after mimetype launcher is fired, try to connect 3 more times
+//  - display success/fail message to user
+frappe.ui.form.qz_connect = function () {
+	return new Promise(function (resolve, reject) {
+		frappe.ui.form.qz_init().then(() => {
+			if (qz.websocket.isActive()) { // if already active, resolve immediately
+				// frappe.show_alert({message: __('QZ Tray Connection Active!'), indicator: 'green'});
+				resolve();
+			} else {
+				// try to connect once before firing the mimetype launcher
+				frappe.show_alert({
+					message: __('Attempting Connection to QZ Tray...'),
+					indicator: 'blue'
+				});
+				qz.websocket.connect().then(() => {
+					frappe.show_alert({
+						message: __('Connected to QZ Tray!'),
+						indicator: 'green'
+					});
+					resolve();
+				}, function retry(err) {
+					if (err.message === 'Unable to establish connection with QZ') {
+						// if a connect was not successful, launch the mimetype, try 3 more times
+						frappe.show_alert({
+							message: __('Attempting to launch QZ Tray...'),
+							indicator: 'blue'
+						}, 14);
+						window.location.assign("qz:launch");
+						qz.websocket.connect({
+							retries: 3,
+							delay: 1
+						}).then(() => {
+							frappe.show_alert({
+								message: __('Connected to QZ Tray!'),
+								indicator: 'green'
+							});
+							resolve();
+						},
+						() => {
+							frappe.throw(__('Error connecting to QZ Tray Application...<br><br> You need to have QZ Tray application installed and running, to use the Raw Print feature.<br><br><a target="_blank" href="https://qz.io/download/">Click here to Download and install QZ Tray</a>.<br> <a target="_blank" href="https://erpnext.com/docs/user/manual/en/setting-up/print/raw-printing">Click here to learn more about Raw Printing</a>.'));
+							reject();
+						});
+					} else {
+						frappe.show_alert({
+							message: 'QZ Tray ' + err.toString(),
+							indicator: 'red'
+						}, 14);
+						reject();
+					}
+				});
+			}
+		});
+	});
+}
+
+frappe.ui.form.qz_init = function () {
+	// Initializing qz tray library
+	return new Promise((resolve) => {
+		if (typeof qz === "object" && typeof qz.version === "string") {
+			// resolve immediately if already Initialized
+			resolve();
+		} else {
+			let qz_required_assets = [
+				"/assets/frappe/node_modules/js-sha256/build/sha256.min.js",
+				"/assets/frappe/node_modules/qz-tray/qz-tray.js"
+			];
+			frappe.require(qz_required_assets,() => {
+				qz.api.setPromiseType(function promise(resolver) {
+					return new Promise(resolver);
+				});
```

Comment:
Just an FYI, we do this internally for you starting with 2.1.1 (as well as the sha256 stuff in the lines following). 🍻